### PR TITLE
[REFACTOR] move FeatureFinderMetaboIdent code from tool to library

### PR DESCRIPTION
**Pull request by timosachsenberg**
# Description

Please include a summary of the change and which issue is fixed.

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.

### CODE REVIEW

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+  String elution_model_;
+
+  // output file (before filtering)
+  String candidates_out_;
+
+  Size debug_level_;
+
+  void updateMembers_() override;
+
+  MSExperiment ms_data_;
```
Comment (oliveralka) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h:196]: Maybe use PeakMap or MSExperiment?

Comment (oliveralka) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h:196]:
```suggestion
  PeakMap ms_data_; ///< input LC-MS data
```

**Comment by oliveralka** [comment]
Looks good IMHO - What is missing in your opinion (ready for review)?

**Comment by timosachsenberg** [comment]
tests & pyOpenMS

**Comment by timosachsenberg** [comment]
rebuild jenkins

**Comment by timosachsenberg** [comment]
Test:
- I think it would be sufficient to add an empty test file for now and refer to the TOPP test similar to https://github.com/OpenMS/OpenMS/blob/develop/src/tests/class_tests/openms/source/FeatureFinderIdentificationAlgorithm_test.cpp
- for pyOpenMS we should check if the bindings already work and maybe also add a test on the pyOpenMS side. We can later extend the binding to e.g. take a pandas dataframe

File: src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd [hunk 2]
```diff
@@ -47,6 +47,11 @@ cdef extern from "<OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMe
...
-  
-
+        String name
+        String formula
+        double mass
+        libcpp_vector[ Int ] charges
+        libcpp_vector[ doubles ] rts
+        libcpp_vector[ doubles ] rt_ranges
+        libcpp_vector[ doubles ] iso_distrib
+
```
Comment (timosachsenberg) [inline src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:57]: probably possible with constructor:
r = FeatureFinderAlgorithmMetaboIdent_Row("caffeine", "C9H3", 200.0, [1], [34.4], null, null)
table = [ FeatureFinderAlgorithmMetaboIdent_Row("caffeine", "C9H3", 200.0, [1], [34.4], null, null), FeatureFinderAlgorithmMetaboIdent_Row("glucose", "C9H3", 200.0, [1], [34.4], null, null) ]

nicer ?:
r = ["caffeine", "C9H3", 200.0, [1], [34.4], null, null]
table = [ "caffeine", "C9H3", 200.0, [1], [34.4], null, null], ["glucose", "C9H3", 200.0, [1], [34.4], null, null] ]

how to get from tsv to table? pyx file needed?

Other question: can we include pandas?
would be probably best solution

Comment (timosachsenberg) [reply src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:57]: Other possibilities:
provide list of columns
 [ ["caffeine", "glucose"], ["C9H3", "C4H2"], [[200.0], [300.0]], [[1,2],[2]],  [[34.4], [45.4]], [null, null], [null, null]]

Comment (timosachsenberg) [reply src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:57]: https://tedboy.github.io/pandas/enhancingperf/enhancingperf1.html
´´´
In [10]: %%cython
   ....: cimport numpy as np
   ....: import numpy as np
   ....: cdef double f_typed(double x) except? -2:
   ....:     return x * (x - 1)
   ....: cpdef double integrate_f_typed(double a, double b, int N):
   ....:     cdef int i
   ....:     cdef double s, dx
   ....:     s = 0
   ....:     dx = (b - a) / N
   ....:     for i in range(N):
   ....:         s += f_typed(a + i * dx)
   ....:     return s * dx
   ....: cpdef np.ndarray[double] apply_integrate_f(np.ndarray col_a, np.ndarray col_b, np.ndarray col_N):
   ....:     assert (col_a.dtype == np.float and col_b.dtype == np.float and col_N.dtype == np.int)
   ....:     cdef Py_ssize_t i, n = len(col_N)
   ....:     assert (len(col_a) == len(col_b) == n)
   ....:     cdef np.ndarray[double] res = np.empty(n)
   ....:     for i in range(len(col_a)):
   ....:         res[i] = integrate_f_typed(col_a[i], col_b[i], col_N[i])
   ....:     return res
´´´

Comment (timosachsenberg) [reply src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:57]: Idee: pandas loads tsv -> converts to lists of lists -> runs pyx function that parses list of lists, builds rows and runs algorithm

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+namespace OpenMS
+{
+class IsotopeDistribution;
+  
+class OPENMS_DLLAPI FeatureFinderAlgorithmMetaboIdent :
+  public DefaultParamHandler
+{
+public:
+  struct OPENMS_DLLAPI FeatureFinderMetaboIdentCompound
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h:57]: add a @brief documentation

File: src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+    std::vector<double> iso_distrib;
+  };
+
+  /// default constructor
+  FeatureFinderAlgorithmMetaboIdent();
+
+  /// @brief perform targeted feature extraction of compounds from @param metaboIdentTable and stores them in @param feature
+  void run(const std::vector<FeatureFinderMetaboIdentCompound>& metaboIdentTable, FeatureMap& features);
+
+  PeakMap& getMSData() { return ms_data_; }
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/TRANSFORMATIONS/FEATUREFINDER/FeatureFinderAlgorithmMetaboIdent.h:92]: add documentation to member functions

File: src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        void run(const libcpp_vector[ FeatureFinderMetaboIdentCompound ] metaboIdentTable, FeatureMap& features) nogil except + #wrap-doc:Run the experiment
+
+        MSExperiment& getChromatograms() nogil except + #wrap-doc:Retrieve chromatograms (empty if run was not executed)
+
+        const TargetedExperiment& getLibrary () nogil except + #wrap-doc:Retrieve the assay library (e.g., to store as TraML, empty if run was not executed)
+        
+        const TransformationDescription& getTransformations() nogil except + #wrap-doc:Retrieve deviations between provided coordinates and extacted ones (e.g., to store as TrafoXML or for plotting)
+
+        size_t getNShared() nogil except + #wrap-doc:Retrieve number of features with shared identifications
```
Comment (timosachsenberg) [inline src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:43]: e.g., copy that comment back so we have it also in the c++ header

File: src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        #   -----
+        #   The algorithms detects quantitative features in MS1 data for a list of targets, typically small molecule/metabolite identifications.
+        #   Internally, it uses algorithms for targeted data analysis from the OpenSWATH pipeline.
+        #   In the simplest case, only CompoundName, SumFormula, Charge and RetentionTime need to be given, all other values may be zero.
+        #   Every combination of compound (mass), RT and charge defines one target for feature detection.
+        #   Output:
+        #   The main output is a feature map of detected features, with annotations in meta data entries.
+        #   Additional outputs are the extracted chromatograms/peak groups, the assay in TraML compatible format, and transformations 
+        #   that contain the error between provided and observed peaks.
+
```
Comment (timosachsenberg) [inline src/pyOpenMS/pxds/FeatureFinderAlgorithmMetaboIdent.pxd:29]: would be great to also have a short usage example here similar to:
https://github.com/OpenMS/OpenMS/blob/develop/src/pyOpenMS/pxds/MSExperiment.pxd#L37-L44

**Review by timosachsenberg — COMMENTED** [review]
looks good. only minor documentation improvements

**Comment by jpfeuffer** [comment]
rebuild jenkins

**Pull request merged**